tests/rptest/tests/many_partitions_test.py

```diff
@@ -0,0 +1,316 @@
+        """
+        Don't do anything with the messages, just consume them to demonstrate
+        that doing so does not exhaust redpanda resources.
+        """
+        def consumer_saw_msgs(consumer):
+            self.logger.info(
+                f"Consumer message_count={consumer.message_count} / {msg_count_per_topic}"
+            )
+            # Tolerate greater-than, because if there were errors during production
+            # there can have been retries.
+            return consumer.message_count >= msg_count_per_topic
+
+        for tn in topic_names:
+            consumer = RpkConsumer(self._ctx,
+                                   self.redpanda,
+                                   tn,
+                                   no_save_msgs=True,
+                                   fetch_max_bytes=BIG_FETCH,
+                                   num_msgs=msg_count_per_topic)
+            consumer.start()
+            wait_until(lambda: consumer_saw_msgs(consumer),
+                       timeout_sec=timeout_per_topic,
+                       backoff_sec=5)
+            consumer.stop()
+            consumer.free()
+
+    @cluster(num_nodes=8)
+    @parametrize(n_partitions=100,
+                 n_topics=1)  # 100 partitions (baseline non-stressed test)
+    @parametrize(n_partitions=int(NODE_MEMORY_MB / 10),
+                 n_topics=10)  # 1 partition per MB spread across 10 topics
+    @parametrize(n_partitions=NODE_MEMORY_MB,
+                 n_topics=1)  # 1 partition per M in one topic
+    def test_many_partitions(self, n_partitions: int, n_topics: int):
+        """
+        Validate that redpanda works with partition counts close to its resource
+        limits.
+
+        This test should evolve over time as we improve efficiency and can reliably
+        run with higher partition counts.  It should roughly track the values we
+        use for topic_memory_per_partition and topic_fds_per_partition.
+
+        * Check topic can be created.
+        * Check leadership election succeeds for all partitions.
+        * Write in enough data such that an unlimited size fetch
+          would exhaust ram (check enforcement of kafka_max_bytes_per_fetch).
+        * Consume all the data from the topic
+
+        * Restart nodes 10 times (check that recovery works, and that the additional
+          log segments created by rolling segments on restart do not cause us
+          to exhaust resources.
+
+        * Run a general produce+consume workload to check that the system remains in
+          a functional state.
+        """
+
+        replication_factor = 3
+
+        # Do a phony allocation of all the non-redpanda nodes, so that we do not
+        # fail the test in CI from having asked for nodes we didn't use.  We're asking
+        # for extra nodes to prevent being run in parallel with other tests that
+        # overcommit the memory on the test machine
+        alloc_nodes = self._ctx.cluster.alloc(ClusterSpec.simple_linux(5))
```

Comment:
Why 5 nodes?

Reply:
This test is asking for mode nodes than it needs in the @cluster decorator, because it needs lots of resources & by default our test running environment oversubscribes containers.  If we don't do this extra alloc, then the test runner will complain that we're not using all our nodes -- the 5 is the delta between how many nodes we're asking for and how many we're really using.

This is very hacky, and should go away once we have a test environment that gives us non-oversubscribed resources.

Reply:
genuinely fill bad for @cluster's inbox